<a name = 'head'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px;; text-align: center; border-top: 2px solid #A523FF; border-bottom: 2px solid #FFB83C; font-size: 28px;' colspan="3">Finance APIs: FXCM</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 17px; font-weight: 525; background-color: #0F793C; padding: 8px; line-height: 1.5; color:white; border-radius: 3px; text-align: center;' colspan="3">|  <a href = 'https://www.fxcm.com/markets/'>FXCM Online</a>  |  <a href = 'https://tradingstation.fxcm.com/'>Account Login</a>  |</font></th>
		</tr>
	</table>
</html>

<html>
	<table>
	<tr style = 'border-top: 2px solid #08CC5B; background-color: #6C07B2; color: white; border-bottom: px solid #08CC5B;' colspan = 3> 
			<td style = 'font-size: 13px; text-align: center;' colspan = 5><b>Evan Marie online</b>:  |  <a href = 'https://www.evanmarie.com/'>EvanMarie.com</a>|  <a href = 'mailto:evanmarie@proton.me'>EvanMarie@Proton.me</a>  |  <a href = 'https://www.linkedin.com/in/evanmarie/'>Linked In</a>  |  <a href = 'https://github.com/EvanMarie'>GitHub</a>   |  <a href = 'https://huggingface.co/EvanMarie'>Hugging Face</a>  |  <a href = 'https://sigmoid.social/@EvanMarie'>Mastadon</a>  |  <a href = 'https://jovian.ai/evanmarie'>Jovian.ai</a>  |  <a href = 'https://www.tiktok.com/@evanmarie_123'>TikTok</a>  |  <a href = 'https://www.codewars.com/users/EvanMarie'>CodeWars</a>  |  <b>Discord ⇨ ✨ EvanMarie ✨#6114</b>  |
			</td></tr>
	</table>
</html>

* Leader in online foreign exchange trading
* Stocks and CFD trading (contract for difference, highly levered financial instrument)
* Cryptocurrencies

<html>
	<table style="width:50%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: center;' colspan="3"><b>Pros and Cons of FXCM</b></font></th>
		</tr>
	</table>
</html>
<font size = 4>Pros:<br></font>
- Completely free with a demo account, which does not allow live trading<br>
- Large datasets available for currencies, commodities, and indexes<br>
- High frequency real-time data<br>
- You can stream the high frequency real-time data<br>
- Bid and ask prices are available<br>
- fxcmpy is a powerful and professional wrapper package for python<br>
	
<br>

<font size = 4>Cons:<br></font>
- Must register and make a demo account
- marketing calls and emails to create a live account
- demo accounts are not available in all countries, for example, the United State
- No data for single stocks and bonds, no fundamentals
- Only provide indexes, currencies, etc
	

In [1]:
from helpers_mardi_gras import *
import fxcmpy
import_all()
plt.style.use('mardi_gras.mplstyle')

In [2]:
%%html
<style>
a:link {color: #FFB83C !important; font-weight: 600 !important;}
a:visited {color: #FFB83C !important; font-weight: 600 !important;}
</style>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 20px; font-weight: 525; background-color: #08CC5B; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: center;' colspan="3">FXCM: First Steps</font></th>
		</tr>
	</table>
</html>

In [3]:
api_key = ''
api = fxcmpy.fxcmpy(access_token = api_key, log_level = 'error')

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: left;' colspan="3"><b># api is an fxcmpy object</b></font></th>
		</tr>
	</table>
</html>

In [4]:
api

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: left;' colspan="3"><b><i># api.get_instruments()</b></i> - trading instruments</font></th>
		</tr>
	</table>
</html>

In [5]:
pretty('api.get_instruments')
help(api.get_instruments)


Help on method get_instruments in module fxcmpy.fxcmpy:

get_instruments() method of fxcmpy.fxcmpy.fxcmpy instance
    Return the tradeable instruments of FXCM as a list.



In [6]:
trading_instruments = pd.Series(api.get_instruments())

In [7]:
see(trading_instruments.sample(10), '10 samples of trading instrumentsv')

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: left;' colspan="3"><b><i>api.close()</i></b>- disconnect from the server</font></th>
		</tr>
	</table>
</html>

In [8]:
# api.close()

<a name = 'historical'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>Downloading Historical Data</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 16px; font-weight: 525; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3">-<i> get_candles()</i></font></th>
		</tr>
	</table>
</html>

In [9]:
pretty('api.get_candles()')
help(api.get_candles)


Help on method get_candles in module fxcmpy.fxcmpy:

get_candles(instrument='', offer_id=None, period='H1', number=10, start=None, end=None, with_index=True, columns=[], stop=None) method of fxcmpy.fxcmpy.fxcmpy instance
    Return historical data from the fxcm database as pandas.DataFrame.
    
    Arguments:
    
    instrument: string (default ''):
        the instrument for which data is requested. For a list of all
        available instruments for historical data, use
        get_instruments_for_candles().
        If the value is equal to '' (default), offer_id must have a value. 
        If both, instrument and offer_id are given, the value of instrument 
        is used.
    
    offer_id: integer (default None): 
        the id of the instrument for which data is requested as given in the
        offer trading table as given by get_offers(). If offer_id is equal 
        to None (default), the parameter instrument must have a value. 
        If both, instrument and offer_id a

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: left;' colspan="3"><b><i># hourly data by default</i></b></font></th>
		</tr>
	</table>
</html>

In [10]:
pretty("api.get_candles('EUR/USD')")
api.get_candles('EUR/USD').head(5).style.format(precision = 4)

In [11]:
pretty("api.get_candles('USD/JPY')")
api.get_candles('USD/JPY').head(5).style.format(precision = 4)

In [12]:
pretty("api.get_candles('GBP/USD')")
api.get_candles('GBP/USD').head(5).style.format(precision = 4)

<a name = 'intraday'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>Intraday Frequency</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3">- <b><i>period</b></i>: string, the granularity of the data<br>- Must be one of: 'm1', 'm5', 'm15', 'm30', 'H1', 'H2', 'H3', 'H4', 'H6', 'H8', 'D1', 'W1', or 'M1'<br>- <b><i>number</b></i>: integer (default 10),
        the number of candles to receive.</font></th>
		</tr>
	</table>
</html>

In [13]:
# api = fxcmpy.fxcmpy(access_token = api_key, log_level = 'error')

In [14]:
pretty("api.get_candles('EUR/USD', period = 'D1'")
api.get_candles('EUR/USD', period = 'D1').head(5).style.format(precision = 4)

In [15]:
pretty("api.get_candles('EUR/USD', period = 'm1')")
api.get_candles('EUR/USD', period = 'm1').head(5).style.format(precision = 4)

In [16]:
import random
instrument = random.choice(trading_instruments)

In [17]:
pretty(f"api.get_candles('Copper', period = 'H3')")
api.get_candles('Copper', period = 'H3').head(5).style.format(precision = 4)

<a name = 'time_period'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>Defining a Time Period</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3">- by default, FXCM returns the 10 most recent time stamps<br>- this section will cover the ways to specify other parameters
		</font></th>
		</tr>
	</table>
</html>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3">Defining time with <i><b>number </b></i></font></th>
		</tr>
	</table>
</html>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: left;' colspan="3"><b><i># daily data, number = 10000 </i></b>(maxiumum number possible)</font></th>
		</tr>
	</table>
</html>

In [18]:
api.get_candles('EUR/USD', 
				period = 'D1', 
				number = 10000)

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2001-09-09 21:00:00     0.91      0.90     0.91    0.90     0.91      0.90   
2001-09-10 21:00:00     0.90      0.91     0.91    0.90     0.90      0.91   
2001-09-11 21:00:00     0.91      0.91     0.92    0.90     0.91      0.91   
2001-09-12 21:00:00     0.91      0.91     0.91    0.90     0.91      0.91   
2001-09-13 21:00:00     0.91      0.92     0.92    0.91     0.91      0.92   
...                      ...       ...      ...     ...      ...       ...   
2023-02-05 22:00:00     1.08      1.07     1.08    1.07     1.08      1.07   
2023-02-06 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   
2023-02-07 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   
2023-02-08 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   
2023-02-09 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   

                     askhigh  asklow  tickqty  
date                                           
2001-09-09 21:00:00     0.91    0.90        0  
2001-09-10 21:00:00     0.92    0.90        0  
2001-09-11 21:00:00     0.92    0.90        0  
2001-09-12 21:00:00     0.91    0.91        0  
2001-09-13 21:00:00     0.93    0.91        0  
...                      ...     ...      ...  
2023-02-05 22:00:00     1.08    1.07   293847  
2023-02-06 22:00:00     1.08    1.07   406035  
2023-02-07 22:00:00     1.08    1.07   267334  
2023-02-08 22:00:00     1.08    1.07   285708  
2023-02-09 22:00:00     1.08    1.07   316647  

[6213 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #FFB83C; padding: 5px 5px 3px 10px; line-height: 1.5; text-align: left; color:black;border-radius: 3px;; text-align: left;' colspan="3"><b><i>^ passed 10,000, but received 6,213, because that is the most available</i></b></font></th>
		</tr>
	</table>
</html>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3">Defining time with <i><b>start & end </b></i>(daily data)</font></th>
		</tr>
	</table>
</html>

In [19]:
api.get_candles('EUR/USD', 
				period = 'D1', 
				start = '2001-01-01', 
				end = '2023-02-10')

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2001-09-09 21:00:00     0.91      0.90     0.91    0.90     0.91      0.90   
2001-09-10 21:00:00     0.90      0.91     0.91    0.90     0.90      0.91   
2001-09-11 21:00:00     0.91      0.91     0.92    0.90     0.91      0.91   
2001-09-12 21:00:00     0.91      0.91     0.91    0.90     0.91      0.91   
2001-09-13 21:00:00     0.91      0.92     0.92    0.91     0.91      0.92   
...                      ...       ...      ...     ...      ...       ...   
2023-02-05 22:00:00     1.08      1.07     1.08    1.07     1.08      1.07   
2023-02-06 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   
2023-02-07 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   
2023-02-08 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   
2023-02-09 22:00:00     1.07      1.07     1.08    1.07     1.07      1.07   

                     askhigh  asklow  tickqty  
date                                           
2001-09-09 21:00:00     0.91    0.90        0  
2001-09-10 21:00:00     0.92    0.90        0  
2001-09-11 21:00:00     0.92    0.90        0  
2001-09-12 21:00:00     0.91    0.91        0  
2001-09-13 21:00:00     0.93    0.91        0  
...                      ...     ...      ...  
2023-02-05 22:00:00     1.08    1.07   293847  
2023-02-06 22:00:00     1.08    1.07   406035  
2023-02-07 22:00:00     1.08    1.07   267334  
2023-02-08 22:00:00     1.08    1.07   285708  
2023-02-09 22:00:00     1.08    1.07   316647  

[6213 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3">Getting <i><b>hourly </b></i>data</font></th>
		</tr>
	</table>
</html>

In [20]:
api.get_candles('EUR/USD', 
				period = 'H1', 
				number = 10000)

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2021-07-16 14:00:00     1.18      1.18     1.18    1.18     1.18      1.18   
2021-07-16 15:00:00     1.18      1.18     1.18    1.18     1.18      1.18   
2021-07-16 16:00:00     1.18      1.18     1.18    1.18     1.18      1.18   
2021-07-16 17:00:00     1.18      1.18     1.18    1.18     1.18      1.18   
2021-07-16 18:00:00     1.18      1.18     1.18    1.18     1.18      1.18   
...                      ...       ...      ...     ...      ...       ...   
2023-02-10 17:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 18:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 19:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 20:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 21:00:00     1.07      1.07     1.07    1.07     1.07      1.07   

                     askhigh  asklow  tickqty  
date                                           
2021-07-16 14:00:00     1.18    1.18    14956  
2021-07-16 15:00:00     1.18    1.18     9403  
2021-07-16 16:00:00     1.18    1.18     4586  
2021-07-16 17:00:00     1.18    1.18     4946  
2021-07-16 18:00:00     1.18    1.18     6476  
...                      ...     ...      ...  
2023-02-10 17:00:00     1.07    1.07    12466  
2023-02-10 18:00:00     1.07    1.07     7181  
2023-02-10 19:00:00     1.07    1.07     7876  
2023-02-10 20:00:00     1.07    1.07     6375  
2023-02-10 21:00:00     1.07    1.07     2782  

[10000 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Getting hourly data with start and end</b><br>- was able to get more than 10,000<br>- this is the exact date range it would allow, no more<br>- this is a trial and error process, how to maximize the time period</th>
		</tr>
	</table>
</html>

In [21]:
api.get_candles('EUR/USD', 
				period = 'H1', 
				start = '2021-05-28', 
				end = '2023-02-10')

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2021-05-28 00:00:00     1.22      1.22     1.22    1.22     1.22      1.22   
2021-05-28 01:00:00     1.22      1.22     1.22    1.22     1.22      1.22   
2021-05-28 02:00:00     1.22      1.22     1.22    1.22     1.22      1.22   
2021-05-28 03:00:00     1.22      1.22     1.22    1.22     1.22      1.22   
2021-05-28 04:00:00     1.22      1.22     1.22    1.22     1.22      1.22   
...                      ...       ...      ...     ...      ...       ...   
2023-02-09 20:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-09 21:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-09 22:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-09 23:00:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 00:00:00     1.07      1.07     1.07    1.07     1.07      1.07   

                     askhigh  asklow  tickqty  
date                                           
2021-05-28 00:00:00     1.22    1.22     5405  
2021-05-28 01:00:00     1.22    1.22     6388  
2021-05-28 02:00:00     1.22    1.22     3819  
2021-05-28 03:00:00     1.22    1.22     4609  
2021-05-28 04:00:00     1.22    1.22     2541  
...                      ...     ...      ...  
2023-02-09 20:00:00     1.07    1.07    12351  
2023-02-09 21:00:00     1.07    1.07     3114  
2023-02-09 22:00:00     1.07    1.07     2487  
2023-02-09 23:00:00     1.07    1.07     2019  
2023-02-10 00:00:00     1.07    1.07     6311  

[10844 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Getting data by the minute</b></th>
		</tr>
	</table>
</html>

In [22]:
api.get_candles('EUR/USD', 
				period = 'm1', 
				number = 10000)

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2023-02-01 22:24:00     1.10      1.10     1.10    1.10     1.10      1.10   
2023-02-01 22:25:00     1.10      1.10     1.10    1.10     1.10      1.10   
2023-02-01 22:26:00     1.10      1.10     1.10    1.10     1.10      1.10   
2023-02-01 22:27:00     1.10      1.10     1.10    1.10     1.10      1.10   
2023-02-01 22:28:00     1.10      1.10     1.10    1.10     1.10      1.10   
...                      ...       ...      ...     ...      ...       ...   
2023-02-10 21:53:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 21:54:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 21:55:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 21:56:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 21:57:00     1.07      1.07     1.07    1.07     1.07      1.07   

                     askhigh  asklow  tickqty  
date                                           
2023-02-01 22:24:00     1.10    1.10        4  
2023-02-01 22:25:00     1.10    1.10        4  
2023-02-01 22:26:00     1.10    1.10        6  
2023-02-01 22:27:00     1.10    1.10        4  
2023-02-01 22:28:00     1.10    1.10        2  
...                      ...     ...      ...  
2023-02-10 21:53:00     1.07    1.07       33  
2023-02-10 21:54:00     1.07    1.07       35  
2023-02-10 21:55:00     1.07    1.07      102  
2023-02-10 21:56:00     1.07    1.07      122  
2023-02-10 21:57:00     1.07    1.07       72  

[10000 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Getting data by the minute with start and end</b><br>- was able to get close to 16,000 records<br>- this is again the exact date range it would allow, no more<br>- this is more data than yfinance or Alpha Vantage allowed</th>
		</tr>
	</table>
</html>

In [23]:
api.get_candles('EUR/USD', 
				period = 'm1', 
				start = '2023-01-26', 
				end = '2023-02-10')

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2023-01-26 00:00:00     1.09      1.09     1.09    1.09     1.09      1.09   
2023-01-26 00:01:00     1.09      1.09     1.09    1.09     1.09      1.09   
2023-01-26 00:02:00     1.09      1.09     1.09    1.09     1.09      1.09   
2023-01-26 00:03:00     1.09      1.09     1.09    1.09     1.09      1.09   
2023-01-26 00:04:00     1.09      1.09     1.09    1.09     1.09      1.09   
...                      ...       ...      ...     ...      ...       ...   
2023-02-09 23:56:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-09 23:57:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-09 23:58:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-09 23:59:00     1.07      1.07     1.07    1.07     1.07      1.07   
2023-02-10 00:00:00     1.07      1.07     1.07    1.07     1.07      1.07   

                     askhigh  asklow  tickqty  
date                                           
2023-01-26 00:00:00     1.09    1.09       73  
2023-01-26 00:01:00     1.09    1.09       93  
2023-01-26 00:02:00     1.09    1.09      149  
2023-01-26 00:03:00     1.09    1.09       65  
2023-01-26 00:04:00     1.09    1.09       68  
...                      ...     ...      ...  
2023-02-09 23:56:00     1.07    1.07       74  
2023-02-09 23:57:00     1.07    1.07       21  
2023-02-09 23:58:00     1.07    1.07      134  
2023-02-09 23:59:00     1.07    1.07       53  
2023-02-10 00:00:00     1.07    1.07      172  

[15804 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Selecting specific columns to import</b></th>
		</tr>
	</table>
</html>

In [24]:
api.get_candles('EUR/USD', 
				period = 'D1', 
				start = '2001-01-01', 
				end = '2023-02-10',
			   columns = ['bidclose', 'askclose'])

bidclose  askclose
date                                   
2001-09-09 21:00:00      0.90      0.90
2001-09-10 21:00:00      0.91      0.91
2001-09-11 21:00:00      0.91      0.91
2001-09-12 21:00:00      0.91      0.91
2001-09-13 21:00:00      0.92      0.92
...                       ...       ...
2023-02-05 22:00:00      1.07      1.07
2023-02-06 22:00:00      1.07      1.07
2023-02-07 22:00:00      1.07      1.07
2023-02-08 22:00:00      1.07      1.07
2023-02-09 22:00:00      1.07      1.07

[6213 rows x 2 columns]

<a name = 'indexes'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>Historical Prices for Stock Indexes</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3"><center><img src = 'http://www.evanmarie.com/content/files/notebook_images/fxcm02.png' width = 500><br><img src = 'http://www.evanmarie.com/content/files/notebook_images/fxcm01.png' width = 500></center></font></th>
		</tr>
	</table>
</html>

In [25]:
instruments = pd.Series(list(api.get_instruments()))
index_mask = instruments.str.contains(r'\d')
indexes_available = pd.Series(instruments.loc[index_mask]).reset_index()

In [26]:
head_tail_horz(indexes_available, 12, 'Indexes Available on FXCM')

head(12) 
 
 
   
 index 
 0 
 
 
 
 
 0 
 11 
 EUSTX50 
 
 
 1 
 29 
 US2000 
 
 
 2 
 153 
 DB1.de 
 
 
 3 
 163 
 TUI1.de 
 
 
 4 
 223 
 A2M.au 
 
 
 5 
 226 
 Z1P.au 
 
 
 6 
 300 
 TOM2.nl 
 
 
 7 
 302 
 5USNote 
 
 
 8 
 303 
 10USNote 
 
 
 9 
 304 
 2USNote 
 
 
 10 
 309 
 FED30D 
 
 
 11 
 311 
 EURIBOR3M 
 
 
 
   
 
 tail(12) 
 
 
   
 index 
 0 
 
 
 
 
 12 
 312 
 SONIA3M 
 
 
 13 
 322 
 AUS200 
 
 
 14 
 323 
 ESP35 
 
 
 15 
 324 
 FRA40 
 
 
 16 
 325 
 GER30 
 
 
 17 
 326 
 HKG33 
 
 
 18 
 327 
 JPN225 
 
 
 19 
 328 
 NAS100 
 
 
 20 
 331 
 SPX500 
 
 
 21 
 332 
 UK100 
 
 
 22 
 333 
 US30 
 
 
 23 
 336 
 CHN50

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>SPX500</b>: daily data, last 10,000 days<br>- returns price index and not total return index<br>- this is the more commonly used index<br>- price return - just the change in price<br>- total return accounts for dividends and the reinvestment thereof</th>
		</tr>
	</table>
</html>

In [27]:
api.get_candles('SPX500', 
				period = 'D1', 
				number = 10000)

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2001-09-10 00:00:00 1,085.78  1,092.54 1,096.94 1,073.15 1,085.78  1,092.54   
2001-09-17 00:00:00 1,092.54  1,038.77 1,092.54 1,037.46 1,092.54  1,038.77   
2001-09-18 00:00:00 1,038.77  1,032.74 1,046.42 1,029.25 1,038.77  1,032.74   
2001-09-19 00:00:00 1,032.74  1,016.10 1,038.91   984.62 1,032.74  1,016.10   
2001-09-20 00:00:00 1,016.10    984.54 1,016.10   984.49 1,016.10    984.54   
...                      ...       ...      ...      ...      ...       ...   
2023-02-05 22:00:00 4,124.12  4,112.85 4,132.61 4,093.90 4,124.72  4,113.45   
2023-02-06 22:00:00 4,114.08  4,156.83 4,178.52 4,088.76 4,114.68  4,157.23   
2023-02-07 22:00:00 4,157.61  4,123.33 4,169.75 4,111.91 4,158.21  4,123.73   
2023-02-08 22:00:00 4,122.78  4,088.38 4,161.05 4,069.08 4,123.38  4,088.78   
2023-02-09 22:00:00 4,088.54  4,089.93 4,094.96 4,051.16 4,089.14  4,090.53   

                     askhigh   asklow  tickqty  
date                                            
2001-09-10 00:00:00 1,096.94 1,073.15        0  
2001-09-17 00:00:00 1,092.54 1,037.46        0  
2001-09-18 00:00:00 1,046.42 1,029.25        0  
2001-09-19 00:00:00 1,038.91   984.62        0  
2001-09-20 00:00:00 1,016.10   984.49        0  
...                      ...      ...      ...  
2023-02-05 22:00:00 4,133.21 4,094.40   192597  
2023-02-06 22:00:00 4,179.02 4,089.26   336170  
2023-02-07 22:00:00 4,170.35 4,112.41   194720  
2023-02-08 22:00:00 4,161.65 4,069.58   209966  
2023-02-09 22:00:00 4,095.46 4,051.76   217261  

[5533 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Dow Jones Industrial Average</b>: daily data, last 10,000 days</th>
		</tr>
	</table>
</html>

In [28]:
api.get_candles('US30', 
				period = 'D1', 
				number = 10000)

bidopen  bidclose   bidhigh    bidlow   askopen  \
date                                                                    
2001-09-10 00:00:00  9,603.36  9,605.51  9,671.80  9,493.55  9,603.36   
2001-09-17 00:00:00  9,580.32  8,920.70  9,580.32  8,883.40  9,580.32   
2001-09-18 00:00:00  8,922.70  8,903.40  9,022.06  8,861.05  8,922.70   
2001-09-19 00:00:00  8,903.54  8,759.13  8,945.47  8,480.21  8,903.54   
2001-09-20 00:00:00  8,748.61  8,376.21  8,748.82  8,375.72  8,748.61   
...                       ...       ...       ...       ...       ...   
2023-02-05 22:00:00 33,843.20 33,871.70 33,967.90 33,643.10 33,854.70   
2023-02-06 22:00:00 33,878.70 34,106.20 34,244.90 33,624.40 33,884.20   
2023-02-07 22:00:00 34,111.80 34,010.30 34,176.70 33,886.00 34,118.40   
2023-02-08 22:00:00 33,998.90 33,712.10 34,245.20 33,586.20 34,003.70   
2023-02-09 22:00:00 33,720.10 33,864.60 33,885.90 33,504.10 33,724.90   

                     askclose   askhigh    asklow  tickqty  
date                                                        
2001-09-10 00:00:00  9,605.51  9,671.80  9,493.55        0  
2001-09-17 00:00:00  8,920.70  9,580.32  8,883.40        0  
2001-09-18 00:00:00  8,903.40  9,022.06  8,861.05        0  
2001-09-19 00:00:00  8,759.13  8,945.47  8,480.21        0  
2001-09-20 00:00:00  8,376.21  8,748.82  8,375.72        0  
...                       ...       ...       ...      ...  
2023-02-05 22:00:00 33,873.20 33,970.10 33,646.90   457177  
2023-02-06 22:00:00 34,107.70 34,247.10 33,626.60   680590  
2023-02-07 22:00:00 34,011.80 34,181.50 33,888.20   481016  
2023-02-08 22:00:00 33,713.60 34,247.40 33,588.40   471943  
2023-02-09 22:00:00 33,868.40 33,888.10 33,507.90   506106  

[5540 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>GER30</b>: hourly data, May 28, 2021 to current<br>- this is a total return index, not price return</th>
		</tr>
	</table>
</html>

In [29]:
api.get_candles('GER30', 
				period = 'H1', 
				start = '2021-05-28', 
				end = '2023-02-10')

bidopen  bidclose   bidhigh    bidlow   askopen  \
date                                                                    
2021-05-28 00:00:00 15,428.93 15,462.56 15,471.50 15,458.56 15,430.03   
2021-05-28 01:00:00 15,462.56 15,469.50 15,475.74 15,460.56 15,469.07   
2021-05-28 02:00:00 15,469.50 15,465.67 15,470.74 15,456.56 15,470.80   
2021-05-28 03:00:00 15,465.67 15,470.67 15,471.83 15,464.56 15,466.97   
2021-05-28 04:00:00 15,470.67 15,478.83 15,479.74 15,461.45 15,471.97   
...                       ...       ...       ...       ...       ...   
2023-02-09 17:00:00 15,513.50 15,501.60 15,534.60 15,469.50 15,515.50   
2023-02-09 18:00:00 15,501.00 15,481.50 15,509.50 15,470.00 15,503.00   
2023-02-09 19:00:00 15,481.00 15,450.50 15,487.50 15,444.50 15,483.00   
2023-02-09 20:00:00 15,450.10 15,433.80 15,450.60 15,403.50 15,452.10   
2023-02-10 00:00:00 15,396.40 15,412.40 15,415.40 15,395.40 15,400.40   

                     askclose   askhigh    asklow  tickqty  
date                                                        
2021-05-28 00:00:00 15,469.07 15,473.07 15,460.97      320  
2021-05-28 01:00:00 15,470.80 15,477.04 15,463.47      701  
2021-05-28 02:00:00 15,466.97 15,473.07 15,459.07      583  
2021-05-28 03:00:00 15,471.97 15,474.07 15,466.47      521  
2021-05-28 04:00:00 15,480.13 15,481.04 15,462.75      413  
...                       ...       ...       ...      ...  
2023-02-09 17:00:00 15,503.60 15,536.60 15,471.50     7646  
2023-02-09 18:00:00 15,483.50 15,511.50 15,472.00    11274  
2023-02-09 19:00:00 15,452.50 15,489.50 15,446.50     8850  
2023-02-09 20:00:00 15,435.30 15,452.60 15,405.50     9611  
2023-02-10 00:00:00 15,416.40 15,419.40 15,399.40      332  

[8920 rows x 9 columns]

<a name = 'commodities'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>Historical Prices for Commodities</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3">- Oil, Gold, Silver, Copper, Wheat, Corn<br>- FXCM is the best place for commodities information</font></th>
		</tr>
	</table>
</html>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>US Oil</b>: daily, max results</th>
		</tr>
	</table>
</html>

In [30]:
api.get_candles('USOil', 
				period = 'D1', 
				number = 10000)

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2001-09-10 00:00:00    27.95     27.63    28.00   27.57    27.95     27.63   
2001-09-11 00:00:00    27.65     27.77    27.77   27.53    27.65     27.77   
2001-09-17 00:00:00    29.00     28.81    29.25   28.30    29.00     28.81   
2001-09-18 00:00:00    28.80     27.70    28.86   27.60    28.80     27.70   
2001-09-19 00:00:00    27.67     26.72    27.95   25.80    27.67     26.72   
...                      ...       ...      ...     ...      ...       ...   
2023-02-05 22:00:00    73.22     74.42    74.50   72.23    73.26     74.45   
2023-02-06 22:00:00    74.53     77.57    77.57   74.33    74.59     77.60   
2023-02-07 22:00:00    77.45     78.47    78.55   77.07    77.51     78.50   
2023-02-08 22:00:00    78.42     77.64    78.83   76.49    78.46     77.67   
2023-02-09 22:00:00    77.66     79.81    80.31   77.45    77.70     79.84   

                     askhigh  asklow  tickqty  
date                                           
2001-09-10 00:00:00    28.00   27.57        0  
2001-09-11 00:00:00    27.77   27.53        0  
2001-09-17 00:00:00    29.25   28.30        0  
2001-09-18 00:00:00    28.86   27.60        0  
2001-09-19 00:00:00    27.95   25.80        0  
...                      ...     ...      ...  
2023-02-05 22:00:00    74.54   72.27   132901  
2023-02-06 22:00:00    77.61   74.37   140385  
2023-02-07 22:00:00    78.59   77.11   118053  
2023-02-08 22:00:00    78.86   76.53   113971  
2023-02-09 22:00:00    80.35   77.49   133209  

[5525 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Gold in USD</b>: daily data, max results</th>
		</tr>
	</table>
</html>

In [31]:
api.get_candles('XAU/USD', 
				period = 'D1',
				number = 10000)

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2001-09-09 21:00:00   273.25    271.75   273.60   270.95   273.25    271.75   
2001-09-10 21:00:00   271.75    286.25   292.50   271.30   271.75    286.25   
2001-09-11 21:00:00   286.75    278.25   292.50   276.75   286.75    278.25   
2001-09-12 21:00:00   278.25    281.25   282.00   278.00   278.25    281.25   
2001-09-13 21:00:00   281.25    285.55   290.00   279.00   281.25    285.55   
...                      ...       ...      ...      ...      ...       ...   
2023-02-05 22:00:00 1,860.83  1,867.24 1,881.23 1,860.36 1,861.89  1,867.54   
2023-02-06 22:00:00 1,866.84  1,871.79 1,884.31 1,864.85 1,868.56  1,872.09   
2023-02-07 22:00:00 1,872.39  1,875.35 1,886.23 1,868.91 1,873.94  1,875.65   
2023-02-08 22:00:00 1,875.36  1,861.70 1,890.17 1,858.78 1,875.98  1,862.00   
2023-02-09 22:00:00 1,861.18  1,865.26 1,872.07 1,852.59 1,862.72  1,865.91   

                     askhigh   asklow  tickqty  
date                                            
2001-09-09 21:00:00   273.60   270.95        0  
2001-09-10 21:00:00   292.50   271.30        0  
2001-09-11 21:00:00   292.50   276.75        0  
2001-09-12 21:00:00   282.00   278.00        0  
2001-09-13 21:00:00   290.00   279.00        0  
...                      ...      ...      ...  
2023-02-05 22:00:00 1,881.55 1,861.36   439702  
2023-02-06 22:00:00 1,884.68 1,865.22   484418  
2023-02-07 22:00:00 1,886.59 1,869.28   359629  
2023-02-08 22:00:00 1,890.53 1,859.11   375300  
2023-02-09 22:00:00 1,872.45 1,852.98   409643  

[5560 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Siver in USD</b>: daily data, max results</th>
		</tr>
	</table>
</html>

In [32]:
api.get_candles('XAG/USD', 
				period = 'D1',
				number = 10000)

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
1950-02-27 21:00:00     0.73      0.73     0.73    0.73     0.73      0.73   
1950-03-30 21:00:00     0.73      0.73     0.73    0.73     0.73      0.73   
1950-04-27 21:00:00     0.72      0.72     0.72    0.72     0.72      0.72   
1950-05-30 21:00:00     0.73      0.73     0.73    0.73     0.73      0.73   
1950-09-28 21:00:00     0.73      0.73     0.73    0.73     0.73      0.73   
...                      ...       ...      ...     ...      ...       ...   
2023-02-05 22:00:00    22.22     22.07    22.60   22.07    22.27     22.34   
2023-02-06 22:00:00    22.05     22.16    22.44   22.03    22.36     22.20   
2023-02-07 22:00:00    22.11     22.27    22.52   22.11    22.47     22.37   
2023-02-08 22:00:00    22.05     21.96    22.58   21.89    22.38     22.00   
2023-02-09 22:00:00    21.94     21.98    22.29   21.64    22.06     22.04   

                     askhigh  asklow  tickqty  
date                                           
1950-02-27 21:00:00     0.73    0.73        0  
1950-03-30 21:00:00     0.73    0.73        0  
1950-04-27 21:00:00     0.72    0.72        0  
1950-05-30 21:00:00     0.73    0.73        0  
1950-09-28 21:00:00     0.73    0.73        0  
...                      ...     ...      ...  
2023-02-05 22:00:00    22.63   22.16   126839  
2023-02-06 22:00:00    22.48   22.07   155153  
2023-02-07 22:00:00    22.55   22.20   108466  
2023-02-08 22:00:00    22.61   21.93   118493  
2023-02-09 22:00:00    22.33   21.85   134826  

[5601 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Natural Gas</b>: hourly data, max results</th>
		</tr>
	</table>
</html>

In [33]:
api.get_candles('NGAS', 
				period = 'H1',
				number = 10000)

bidopen  bidclose  bidhigh  bidlow  askopen  askclose  \
date                                                                         
2021-06-04 12:00:00     3.06      3.04     3.07    3.04     3.07      3.06   
2021-06-04 13:00:00     3.04      3.04     3.06    3.03     3.06      3.05   
2021-06-04 14:00:00     3.04      3.04     3.05    3.02     3.05      3.05   
2021-06-04 15:00:00     3.04      3.04     3.05    3.03     3.05      3.05   
2021-06-04 16:00:00     3.04      3.10     3.11    3.04     3.05      3.12   
...                      ...       ...      ...     ...      ...       ...   
2023-02-10 17:00:00     2.50      2.51     2.55    2.49     2.51      2.52   
2023-02-10 18:00:00     2.51      2.48     2.52    2.48     2.52      2.49   
2023-02-10 19:00:00     2.48      2.54     2.54    2.47     2.49      2.55   
2023-02-10 20:00:00     2.54      2.57     2.58    2.54     2.55      2.58   
2023-02-10 21:00:00     2.57      2.57     2.58    2.55     2.58      2.58   

                     askhigh  asklow  tickqty  
date                                           
2021-06-04 12:00:00     3.08    3.06     1214  
2021-06-04 13:00:00     3.07    3.04     1464  
2021-06-04 14:00:00     3.06    3.03      984  
2021-06-04 15:00:00     3.06    3.04      784  
2021-06-04 16:00:00     3.13    3.05     2243  
...                      ...     ...      ...  
2023-02-10 17:00:00     2.56    2.50     2761  
2023-02-10 18:00:00     2.53    2.49     1673  
2023-02-10 19:00:00     2.56    2.48     2900  
2023-02-10 20:00:00     2.59    2.55     1342  
2023-02-10 21:00:00     2.59    2.56      640  

[10000 rows x 9 columns]

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Soy Futures</b>: hourly data, max results</th>
		</tr>
	</table>
</html>

In [34]:
api.get_candles('SOYF', 
				period = 'H1',
				number = 10000)

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2020-12-30 14:00:00 1,289.12  1,292.39 1,294.68 1,286.11 1,290.05  1,293.50   
2020-12-30 15:00:00 1,292.39  1,295.60 1,297.93 1,289.84 1,293.50  1,296.33   
2020-12-30 16:00:00 1,295.60  1,297.48 1,304.25 1,294.06 1,296.33  1,298.20   
2020-12-30 17:00:00 1,297.48  1,293.25 1,297.95 1,289.76 1,298.20  1,293.99   
2020-12-30 18:00:00 1,293.25  1,303.45 1,303.93 1,292.36 1,293.99  1,304.42   
...                      ...       ...      ...      ...      ...       ...   
2023-02-10 15:00:00 1,520.26  1,529.11 1,530.26 1,516.99 1,521.52  1,530.37   
2023-02-10 16:00:00 1,529.00  1,521.86 1,529.51 1,519.49 1,530.26  1,523.12   
2023-02-10 17:00:00 1,521.88  1,528.88 1,529.26 1,521.01 1,523.14  1,530.14   
2023-02-10 18:00:00 1,529.01  1,531.26 1,532.26 1,527.24 1,530.27  1,532.52   
2023-02-10 19:00:00 1,531.12  1,532.76 1,533.26 1,530.99 1,532.38  1,534.02   

                     askhigh   asklow  tickqty  
date                                            
2020-12-30 14:00:00 1,295.89 1,287.24    12147  
2020-12-30 15:00:00 1,299.09 1,290.87    13485  
2020-12-30 16:00:00 1,305.39 1,295.22    13203  
2020-12-30 17:00:00 1,298.97 1,290.86     9320  
2020-12-30 18:00:00 1,304.89 1,293.53     8445  
...                      ...      ...      ...  
2023-02-10 15:00:00 1,531.52 1,518.25     5645  
2023-02-10 16:00:00 1,530.77 1,520.75     4305  
2023-02-10 17:00:00 1,530.52 1,522.27     3010  
2023-02-10 18:00:00 1,533.52 1,528.50     2371  
2023-02-10 19:00:00 1,534.52 1,532.25     1063  

[10000 rows x 9 columns]

<a name = 'crypto'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>Historical Prices for Cryptocurrencies</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3">- they used to have Ethereum and a few others, but those no longer work</font></th>
		</tr>
	</table>
</html>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Bitcoin</b>: daily data, start and end dates</th>
		</tr>
	</table>
</html>

In [35]:
api.get_candles('BTC/USD', 
				period = 'D1', 
				start = '2002-01-01', 
				end = '2023-02-10')

bidopen  bidclose   bidhigh    bidlow   askopen  \
date                                                                    
2017-12-15 22:00:00 19,900.00 19,900.00 19,900.00 19,900.00 19,901.00   
2017-12-18 22:00:00 19,900.00 19,705.00 20,040.00 19,300.00 19,901.00   
2017-12-19 22:00:00 19,705.00 18,555.00 19,340.00 19,310.00 19,706.00   
2017-12-20 22:00:00 18,555.00 17,275.00 17,300.00 17,275.00 18,556.00   
2017-12-21 22:00:00 17,275.00 15,680.00 16,330.00 15,915.00 17,276.00   
...                       ...       ...       ...       ...       ...   
2023-02-06 22:00:00 22,975.30 22,893.80 23,133.50 22,621.50 23,015.30   
2023-02-07 22:00:00 22,894.30 23,177.80 23,330.50 22,621.50 22,934.30   
2023-02-08 22:00:00 23,178.30 22,939.40 23,430.50 22,654.50 23,218.30   
2023-02-09 22:00:00 22,939.00 21,839.30 22,992.20 21,678.20 22,979.00   
2023-02-10 22:00:00 21,839.80 21,530.10 21,939.10 21,458.40 21,879.80   

                     askclose   askhigh    asklow  tickqty  
date                                                        
2017-12-15 22:00:00 19,901.00 19,901.00 19,901.00        0  
2017-12-18 22:00:00 19,706.00 20,041.00 19,301.00        0  
2017-12-19 22:00:00 18,556.00 19,341.00 19,311.00        0  
2017-12-20 22:00:00 17,276.00 17,301.00 17,276.00        0  
2017-12-21 22:00:00 15,681.00 16,331.00 15,916.00        0  
...                       ...       ...       ...      ...  
2023-02-06 22:00:00 22,933.80 23,173.50 22,661.50    50536  
2023-02-07 22:00:00 23,217.80 23,370.50 22,661.50    68856  
2023-02-08 22:00:00 22,979.40 23,470.50 22,690.00    59518  
2023-02-09 22:00:00 21,879.30 23,032.20 21,718.20    83253  
2023-02-10 22:00:00 21,564.10 21,979.10 21,498.40    67891  

[1345 rows x 9 columns]

<a name = 'real_time'></a>

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>



<a name = 'head'></a>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #A523FF; padding: 5px 5px 3px 5px; line-height: 1.5; color:white;border-radius: 3px;; text-align: center; border-bottom: 1.5px solid white;' colspan="3"><font size = 6>High Frequency Real-Time Data</font></th>
		</tr>
		<tr>
			<th style = 'font-size: 15px; font-weight: 500; background-color: #FFB83C; padding: 8px; line-height: 1.5; color:black; border-radius: 3px; text-align: left;' colspan="3">- FXCM supports high frequency real-time data for some instruments<br>- this is something that usually costs a great deal for a subscription to access</font></th>
		</tr>
	</table>
</html>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b><i>api.subscribe_market_data()</b></i></b></th>
		</tr>
	</table>
</html>

In [36]:
help(api.subscribe_market_data)

Help on method subscribe_market_data in module fxcmpy.fxcmpy:

subscribe_market_data(symbol='', add_callbacks=()) method of fxcmpy.fxcmpy.fxcmpy instance
    Stream the prices of an instrument.
    
    Arguments:
    
    symbol:  string,
        the symbol of the instrument in question as given by
        get_instruments().
    
    add_callbacks: list of callables,
        all methods in that list will be called for every incoming dataset
        of the instrument. Such a method has to accept two positional
        arguments, data and dataframe, say. The first should be a json like
        object with the new price data received by the stream and the
        second should be a Pandas DataFrame with the collected price data
        as given by get_prices().



<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Subscribe to EUR/USD exchange data</b></th>
		</tr>
	</table>
</html>

In [37]:
api.subscribe_market_data('EUR/USD')

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b><i>api.get_subscribed_symbols()</b></i></b> - get the current open subscriptions</th>
		</tr>
	</table>
</html>

In [38]:
help(api.get_subscribed_symbols)

Help on method get_subscribed_symbols in module fxcmpy.fxcmpy:

get_subscribed_symbols() method of fxcmpy.fxcmpy.fxcmpy instance
    Returns a list of symbols for the subscribed instruments.



In [39]:
api.get_subscribed_symbols()

['EUR/USD']

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b><i>api.get_last_price()</i></b> - get the last reported price for a symbol</th>
		</tr>
	</table>
</html>

In [40]:
help(api.get_last_price)

Help on method get_last_price in module fxcmpy.fxcmpy:

get_last_price(symbol) method of fxcmpy.fxcmpy.fxcmpy instance
    Return the last prices of a given subscribed instrument.
    
    Arguments:
    
    symbol: string,
        the symbol of the instrument as given by get_instruments().



In [41]:
pretty('The most recent price data for EUR/USD')
api.get_last_price('EUR/USD')

Bid    1.07
Ask    1.07
High   1.08
Low    1.07
Name: 2023-02-10 21:57:59.544000, dtype: float64

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b><i>api.get_prices()</i></b> - get all available data for a subscription since the subscription was started</th>
		</tr>
	</table>
</html>

In [42]:
help(api.get_prices)

Help on method get_prices in module fxcmpy.fxcmpy:

get_prices(symbol) method of fxcmpy.fxcmpy.fxcmpy instance
    Return the prices of a given subscribed instrument.
    
    Arguments:
    
    symbol: string,
        the symbol of the instrument as given by get_instruments().



In [43]:
api.get_prices('EUR/USD')

Bid  Ask  High  Low
2023-02-10 21:57:59.544 1.07 1.07  1.08 1.07

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3">Subscribing to Bitcoin also, because Euro is not updating</th>
		</tr>
	</table>
</html>

In [44]:
api.subscribe_market_data('BTC/USD')

In [45]:
api.get_last_price('EUR/USD')

Bid    1.07
Ask    1.07
High   1.08
Low    1.07
Name: 2023-02-10 21:57:59.544000, dtype: float64

In [46]:
api.get_prices('BTC/USD')

Bid    Ask      High       Low
2023-02-10 21:58:30.932  21530  21564 23,470.50 21,458.25

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Stream live data:</b></th>
		</tr>
	</table>
</html>

In [49]:
import time
count = 0
pretty('Streaming EUR/USD data:')
while count < 10:
	time.sleep(1)
	print(api.get_last_price('EUR/USD').name, api.get_last_price('EUR/USD').Ask)
	count += 1


2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803
2023-02-10 21:57:59.544000 1.06803


In [50]:
pretty('time.sleep()')
help(time.sleep)


Help on built-in function sleep in module time:

sleep(...)
    sleep(seconds)
    
    Delay execution for a given number of seconds.  The argument may be
    a floating point number for subsecond precision.



<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Function to stream the current ask price</b></th>
		</tr>
	</table>
</html>

In [51]:
def stream_data(ticker, delay = 1, reps = 10):
	
	api.subscribe_market_data(ticker)
	import time
	
	pretty(f'Streaming {ticker} data:')
	count = 0
	while count < reps:
		time.sleep(delay)
		
		price = api.get_last_price(ticker).Ask
		timestamp = api.get_last_price(ticker).name
		pretty(f"{price:,.3f}", timestamp, fontsize = 3)
		
		count += 1
	api.subscribe_market_data(ticker)
	pretty(f'Transmission complete.')

In [52]:
stream_data('EUR/USD', delay = 1, reps = 3)

In [53]:
stream_data('BTC/USD', delay = 1, reps = 3)

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'font-size: 16px; font-weight: 500; background-color: #6C07B2; padding: 8px; line-height: 1.5; color:white; border-radius: 3px;; text-align: left;' colspan="3"><b>Unsubscribe from market data</b></th>
		</tr>
	</table>
</html>

In [54]:
api.get_subscribed_symbols()

['EUR/USD', 'BTC/USD']

In [60]:
stream_data('LCattleF', delay = 1, reps = 3)

In [61]:
api.unsubscribe_market_data('EUR/USD')
api.unsubscribe_market_data('BTC/USD')

In [62]:
api.close()

<hr class="dashed" style = 'color: #08CC5B'>

<html>
	<table style="width:100%;">
		<tr>
			<th style = 'background-color: #6C07B2; padding: 5px 5px 3px 5px; line-height: 1.5; color:white; border-radius: 3px; font-size: 14px; font-weight: 475; text-align: center; border-top: 2px solid #FFB83C; border-bottom: 2px solid #FFB83C;'  colspan="3"> |  <a href="#head">Top</a>  |  <a href="#historical">Historical Data</a>  |  <a href="#intraday">Intraday Frequency</a>   |  <a href="#time_period">Defining Time Period</a>  |  <a href="#indexes">Stock Indexes</a>  |  <a href="#commodities">Commodities</a>    <a href="#crypto">Cryptocurrencies</a>    <a href="#real_time">Real-Time Data</a>    </th>
		</tr>
	</table>
</html>

<hr class="dashed" style = 'color: #08CC5B'>

